File: clone_graph.py


class UndirectedGraphNode:
    def __init__(self, x):
        self.label = x
        self.neighbors = []

class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        return node

if __name__ == "__main__":
    n0 = UndirectedGraphNode(0)
    n1 = UndirectedGraphNode(1)
    n2 = UndirectedGraphNode(2)
    n0.neighbors = [n1,n2]
    n1.neighbors = [n0,n2]
    n2.neighbors = [n0,n1]
    n = Solution().cloneGraph(n0)
    print n

File: 3/clone_graph.cpp
#include "header.h"

struct UndirectedGraphNode {
    int label;
    vector<UndirectedGraphNode *> neighbors;
    UndirectedGraphNode(int x) : label(x) {};
};

UndirectedGraphNode* dfs(UndirectedGraphNode* root, unordered_map<UndirectedGraphNode*,UndirectedGraphNode*>& map){
    if(!root) return NULL;
    if(map.find(root) != map.end())
        return map[root];
    UndirectedGraphNode* root_c = new UndirectedGraphNode(root->label);
    map[root] = root_c;
    for(int i=0; i<root->neighbors.size(); i++)
        root_c->neighbors.push_back(dfs(root->neighbors[i],map));
    return root_c;
}

UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
    if(!node) return NULL;
    unordered_map<UndirectedGraphNode*,UndirectedGraphNode*> map;
    return dfs(node,map);
}

int main(){
    UndirectedGraphNode n0(0);
    UndirectedGraphNode n1(1);
    UndirectedGraphNode n2(2);
    vector<UndirectedGraphNode*> n0_n;
    vector<UndirectedGraphNode*> n1_n;
    vector<UndirectedGraphNode*> n2_n;
    n0_n.push_back(&n1);
    n0_n.push_back(&n2);
    n1_n.push_back(&n2);
    n1_n.push_back(&n0);
    n2_n.push_back(&n0);
    n2_n.push_back(&n1);
    n2_n.push_back(&n2);
    UndirectedGraphNode* res = cloneGraph(&n0);
    return 0;
}
